Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,664 @@
+
+    public void RecomputePerLineDiagnostics() {
+      PerLineDiagnostic = RenderPerLineDiagnostics(this, PerNodeDiagnostic, LinesCount,
+        NumberOfResolutionErrors, Diagnostics);
+    }
+
+    static LineVerificationStatus[] RenderPerLineDiagnostics(
+      VerificationDiagnosticsParams verificationDiagnosticsParams,
+      NodeDiagnostic[] perNodeDiagnostic,
+      int numberOfLines,
+      int numberOfResolutionErrors,
+      Container<Diagnostic> diagnostics
+    ) {
+      var result = new LineVerificationStatus[numberOfLines];
+
+      if (perNodeDiagnostic.Length == 0) {
+        for (var line = 0; line < numberOfLines; line++) {
+          result[line] = LineVerificationStatus.Verified;
+        }
+
+        return result;
+      }
+
+      // Render node content into lines.
+      foreach (var nodeDiagnostic in perNodeDiagnostic) {
+        if (nodeDiagnostic.Filename == verificationDiagnosticsParams.Uri.GetFileSystemPath() ||
+            "untitled:" + nodeDiagnostic.Filename == verificationDiagnosticsParams.Uri) {
+          nodeDiagnostic.RenderInto(result);
+        }
+      }
+
+      // Fill in the missing "Unknown" based on the surrounding content
+      // The filling only takes Verified an Error
+      var previousNotUnknown = LineVerificationStatus.Unknown;
+      var lineDelta = 1;
+      // Two passes so that we can fill gaps based on what happened before AND after
+      for (var line = 0; 0 <= line; line += lineDelta) {
+        if (line == numberOfLines) {
+          lineDelta = -1;
+          previousNotUnknown = LineVerificationStatus.Unknown;
+          continue;
+        }
+        if (previousNotUnknown != LineVerificationStatus.Verified &&
+            previousNotUnknown != LineVerificationStatus.VerifiedObsolete &&
+            previousNotUnknown != LineVerificationStatus.VerifiedVerifying) {
+          previousNotUnknown = LineVerificationStatus.Unknown;
+        }
+        if (result[line] == LineVerificationStatus.Unknown) {
+          result[line] = previousNotUnknown;
+        } else {
+          previousNotUnknown = result[line];
+        }
+      }
+
+      var resolutionErrorRendered = 0;
+      foreach (var diagnostic in diagnostics) {
+        if (resolutionErrorRendered >= numberOfResolutionErrors) {
+          break;
+        }
+        result[diagnostic.Range.Start.Line] = LineVerificationStatus.ResolutionError;
+        resolutionErrorRendered++;
+      }
+
+      var existsErrorRange = false;
+      var existsError = false;
+      foreach (var line in result) {
+        existsErrorRange = existsErrorRange || line == LineVerificationStatus.ErrorRange;
+        existsError = existsError || line == LineVerificationStatus.Error;
+      }
+
+      if (existsErrorRange && !existsError) {
+        existsError = false;
+      }
+
+      return result;
+    }
+  }
+
+
+  public enum VerificationStatus {
+    Unknown = 0,
+    Verified = 200,
+    Inconclusive = 270,
+    Error = 400
+  }
```

Comment:
This is public API and should carefully be agreed on. It is the verification status of every node.
Note that, when rendering linearly, Inconclusive nodes are not displayed (so that we don't distract the user with assertions that were not checked)